File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/teleport/RandomTeleportResult.java
package top.likoslupus.cellulosesz.api.teleport;

import java.util.Optional;

import static java.util.Objects.requireNonNull;

public record RandomTeleportResult(
        RandomTeleportStatus status,
        Optional<CellLocation> location
) {

    public RandomTeleportResult {
        requireNonNull(status, "status");
        requireNonNull(location, "location");
        if (status == RandomTeleportStatus.SUCCESS && location.isEmpty()) {
            throw new IllegalArgumentException("successful result requires a location");
        }
    }

    public static RandomTeleportResult success(CellLocation location) {
        return new RandomTeleportResult(
                RandomTeleportStatus.SUCCESS,
                Optional.of(requireNonNull(location, "location"))
        );
    }

    public static RandomTeleportResult failure(RandomTeleportStatus status) {
        if (status == RandomTeleportStatus.SUCCESS) {
            throw new IllegalArgumentException("failure status required");
        }
        return new RandomTeleportResult(status, Optional.empty());
    }

    public boolean success() {
        return status == RandomTeleportStatus.SUCCESS;
    }

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/teleport/RandomTeleportResult.kt
package top.likoslupus.cellulosesz.api.teleport

@JvmRecord
public data class RandomTeleportResult(
    public val status: RandomTeleportStatus,
    public val location: CellLocation?
) {

    init {
        require(!(status == RandomTeleportStatus.SUCCESS && location == null)) {
            "successful result requires a location"
        }
    }

    public fun success(): Boolean = status == RandomTeleportStatus.SUCCESS

    public companion object {

        @JvmStatic
        public fun success(location: CellLocation): RandomTeleportResult =
            RandomTeleportResult(RandomTeleportStatus.SUCCESS, location)

        @JvmStatic
        public fun failure(status: RandomTeleportStatus): RandomTeleportResult {
            require(status != RandomTeleportStatus.SUCCESS) { "failure status required" }
            return RandomTeleportResult(status, null)
        }

    }

}
